src/transport.rs

```diff
@@ -11,51 +11,648 @@
+};
+use crate::{RegistrationType, UPTransportVsomeip};
+use crate::{
+    TransportCommand, UP_CLIENT_VSOMEIP_FN_TAG_REGISTER_LISTENER_INTERNAL, UP_CLIENT_VSOMEIP_TAG,
+};
 use async_trait::async_trait;
+use cxx::{let_cxx_string, SharedPtr};
+use lazy_static::lazy_static;
+use log::{error, trace, warn};
+use once_cell::sync::Lazy;
+use std::collections::{HashMap, HashSet};
+use std::hash::{DefaultHasher, Hash, Hasher};
+use std::sync::{mpsc, Arc};
+use std::time::Duration;
+use tokio::runtime::Runtime;
+use tokio::sync::oneshot;
+use tokio::sync::Mutex;
+use tokio::sync::RwLock;
+use tokio::task::LocalSet;
+use tokio::time::{timeout, Instant};
+use up_rust::{ComparableListener, UCode, UListener, UMessage, UStatus, UTransport, UUri};
+use vsomeip_proc_macro::generate_message_handler_extern_c_fns;
+use vsomeip_sys::extern_callback_wrappers::MessageHandlerFnPtr;
+use vsomeip_sys::glue::{make_application_wrapper, make_message_wrapper, make_runtime_wrapper};
+use vsomeip_sys::safe_glue::get_pinned_runtime;
+use vsomeip_sys::vsomeip;
 
-use up_rust::{
-    transport::datamodel::UTransport,
-    uprotocol::{UMessage, UStatus, UUri},
-    uuid::builder::UUIDBuilder,
-};
+const UP_CLIENT_VSOMEIP_FN_TAG_REGISTER_LISTENER: &str = "register_listener";
+
+const INTERNAL_FUNCTION_TIMEOUT: u64 = 3;
+
+static RUNTIME: Lazy<Arc<Runtime>> =
+    Lazy::new(|| Arc::new(Runtime::new().expect("Failed to create Tokio runtime")));
+
+fn get_runtime() -> Arc<Runtime> {
+    Arc::clone(&RUNTIME)
+}
+
+const THREAD_NUM: usize = 10;
+
+// Create a separate tokio Runtime for running the callback
+lazy_static! {
+    static ref CB_RUNTIME: Runtime = tokio::runtime::Builder::new_multi_thread()
+        .worker_threads(THREAD_NUM)
+        .enable_all()
+        .build()
+        .expect("Unable to create callback runtime");
+}
+
+lazy_static! {
+    pub(crate) static ref AUTHORITY_NAME: Mutex<String> = Mutex::new(String::new());
+    pub(crate) static ref LISTENER_CLIENT_ID_MAPPING: RwLock<HashMap<usize, ClientId>> =
+        RwLock::new(HashMap::new());
+    pub(crate) static ref CLIENT_ID_APP_MAPPING: RwLock<HashMap<ClientId, String>> =
+        RwLock::new(HashMap::new());
+    pub(crate) static ref UE_REQUEST_CORRELATION: RwLock<HashMap<RequestId, ReqId>> =
+        RwLock::new(HashMap::new());
+    pub(crate) static ref ME_REQUEST_CORRELATION: RwLock<HashMap<ReqId, RequestId>> =
+        RwLock::new(HashMap::new());
+    pub(crate) static ref CLIENT_ID_SESSION_ID_TRACKING: RwLock<HashMap<ClientId, SessionId>> =
+        RwLock::new(HashMap::new());
+}
 
-use crate::UPClientVsomeip;
+type ListenerIdMap = RwLock<HashMap<(UUri, Option<UUri>, ComparableListener), usize>>;
+
+lazy_static! {
+    pub(crate) static ref LISTENER_REGISTRY: RwLock<HashMap<usize, Arc<dyn UListener>>> =
+        RwLock::new(HashMap::new());
+    pub(crate) static ref LISTENER_ID_MAP: ListenerIdMap = RwLock::new(HashMap::new());
+}
+
+generate_message_handler_extern_c_fns!(10000);
```

Comment:
You can check the proc macro crate's definition of generate_message_handler_extern_c_fns. Long story short is that since an extern "C" fn is required by the vsomeip library to register a message handler function, we must pre-create all these extern "C" fns.

Internal to each extern "C" fn they call call_shared_extern_fn function which will then use the map to pick out the correct UListener to trigger.